File: BackEnd/PHP/laravelshop/app/Common/Http/ApiResponse.php
<?php

namespace App\Common\Http;

use Illuminate\Http\JsonResponse;

final class ApiResponse
{
    public static function success(mixed $data = null, array $meta = [], int $status = 200): JsonResponse
    {
        $payload = [
            'success' => true,
            'data' => $data,
        ];

        if ($meta !== []) {
            $payload['meta'] = $meta;
        }

        return response()->json($payload, $status);
    }

    public static function error(string $message, string $code = 'INTERNAL_SERVER_ERROR', int $status = 500, array $details = []): JsonResponse
    {
        $payload = [
            'success' => false,
            'error' => [
                'code' => $code,
                'message' => $message,
            ],
        ];

        if ($details !== []) {
            $payload['error']['details'] = $details;
        }

        return response()->json($payload, $status);
    }
}

File: BackEnd/Java/springshop_maven/src/main/java/com/pbshop/springshop/common/api/ApiResponse.java
package com.pbshop.springshop.common.api;

import java.time.OffsetDateTime;
import java.util.Map;

public record ApiResponse<T>(
    boolean success,
    T data,
    ApiMeta meta,
    ApiError error
) {
    public static <T> ApiResponse<T> success(T data) {
        return new ApiResponse<>(
            true,
            data,
            new ApiMeta(OffsetDateTime.now().toString(), "ko-KR", "KRW"),
            null
        );
    }

    public static ApiResponse<Void> failure(String code, String message) {
        return new ApiResponse<>(
            false,
            null,
            new ApiMeta(OffsetDateTime.now().toString(), "ko-KR", "KRW"),
            new ApiError(code, message, Map.of())
        );
    }

    public record ApiMeta(
        String timestamp,
        String locale,
        String currency
    ) {}

    public record ApiError(
        String code,
        String message,
        Map<String, Object> details
    ) {}
}
